plugins/module_utils/sops.py

```diff
@@ -114,57 +119,90 @@ def f(value, arguments, env):
 class SopsError(Exception):
     ''' Extend Exception class with sops specific information '''
 
-    def __init__(self, filename, exit_code, message, decryption=True):
+    def __init__(self, filename, exit_code, message, decryption=True, operation=None):
+        if operation is None:
+            operation = 'decrypt' if decryption else 'encrypt'
         if exit_code in SOPS_ERROR_CODES:
             exception_name = SOPS_ERROR_CODES[exit_code]
             message = "error with file %s: %s exited with code %d: %s" % (
                 filename, exception_name, exit_code, to_native(message))
         else:
             message = "could not %s file %s; Unknown sops error code: %s; message: %s" % (
-                'decrypt' if decryption else 'encrypt', filename, exit_code, to_native(message))
+                operation, filename, exit_code, to_native(message))
         super(SopsError, self).__init__(message)
 
 
-class Sops():
-    ''' Utility class to perform sops CLI actions '''
+SopsFileStatus = collections.namedtuple('SopsFileStatus', ['encrypted'])
 
-    @staticmethod
-    def _add_options(command, env, get_option_value, options):
+
+class SopsRunner(object):
+    def _add_options(self, command, env, get_option_value, options):
         if get_option_value is None:
             return
         for option, f in options.items():
             v = get_option_value(option)
             if v is not None:
-                f(v, command, env)
+                f(v, command, env, self.version)
+
+    def _debug(self, message):
+        if self.display:
+            self.display.vvvv(message)
+        elif self.module:
+            self.module.debug(message)
+
+    def _warn(self, message):
+        if self.display:
+            self.display.warning(message)
+        elif self.module:
+            self.module.warn(message)
+
+    def __init__(self, binary, module=None, display=None):
+        self.binary = binary
+        self.module = module
+        self.display = display
+
+        self.version = (3, 7, 3)  # if --disable-version-check is not supported, this is version 3.7.3 or older
+        self.version_string = '(before 3.8.0)'
+
+        exit_code, output, err = self._run_command([self.binary, '--version', '--disable-version-check'])
+        if exit_code == 0:
+            m = _SOPS_VERSION.match(output.decode('utf-8'))
+            if m:
+                self.version = int(m.group(1)), int(m.group(2)), int(m.group(3))
+                self.version_string = '%d.%d.%d' % self.version
+                self._debug('SOPS version detected as %s' % (self.version, ))
+            else:
+                self._warn('Cannot extract SOPS version from: %s' % repr(output))
+        else:
+            self._debug('Cannot detect SOPS version efficiently, likely a version before 3.8.0')
```

Comment:
I'd suggest to remove `efficiently` as it suggest that some form of version detection happened, which is not the case.
```suggestion
            self._debug('Cannot detect SOPS version, likely a version before 3.8.0')
```

Reply:
Actually efficicently is the whole point here: only 3.8.0 and later allow to disable the network request done when calling `sops --version`, which makes `sops --version` rather slow (0.3-0.7 seconds vs. 0.01-0.03 seconds runtime for 3.9.0, not sure how slow it is with older versions, since the way the version check worked changed over time).

Reply:
Thank you for the clarification! So that cli flag avoids the roundtrip speeding up the command. Let's merge this! 👍